clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp

```diff
@@ -53,6 +57,173 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
   if (CStyleCastDecl != nullptr) {
     diag(CStyleCastDecl->getLParenLoc(), "C-style cast MUST not be used");
   }
+
+  const auto *ImplicitCastDecl =
+      Result.Nodes.getNodeAs<ImplicitCastExpr>(ImplicitCastId);
+  if (ImplicitCastDecl != nullptr) {
+    const QualType destType = ImplicitCastDecl->getType();
+    const QualType sourceType = ImplicitCastDecl->getSubExpr()->getType();
+    if (ImplicitCastDecl->isPartOfExplicitCast()) {
+      handleStaticCast(Result.Context, sourceType, destType,
+                       ImplicitCastDecl->getBeginLoc());
+    } else {
+      handleImplicitCast(Result.Context, sourceType, destType,
+                         ImplicitCastDecl->getBeginLoc());
+    }
+  }
+
+  const auto *StaticCastDecl =
+      Result.Nodes.getNodeAs<CXXStaticCastExpr>(StaticCastId);
+  if (StaticCastDecl != nullptr) {
+    const QualType destType = StaticCastDecl->getType();
+    const QualType sourceType = StaticCastDecl->getSubExpr()->getType();
+    handleStaticCast(Result.Context, sourceType, destType,
+                     StaticCastDecl->getOperatorLoc());
+  }
+}
+
+bool TypeConversionsCheck::isValidIntegralPromotion(clang::ASTContext *context,
+                                                    QualType sourceType,
+                                                    QualType destType) const {
+  if (sourceType->isIntegerType() && destType->isIntegerType()) {
+    // Check that the size of the destination integer is larger than source
+
+    const unsigned sourceBits = sourceType->isBitIntType()
+                                    ? context->getIntWidth(sourceType)
+                                    : context->getTypeSize(sourceType);
+    const unsigned destBits = destType->isBitIntType()
+                                  ? context->getIntWidth(destType)
+                                  : context->getTypeSize(destType);
+
+    // unsigned to signed conversion is fine if number of bits is strictly
+    // larger than the source, otherwise it may not fit
+    if ((destType->isSignedIntegerType() &&
+         sourceType->isUnsignedIntegerType()) &&
+        (destBits > sourceBits)) {
+      return true;
+    }
+
+    // If they are the same sign, they need to have at least the same number of
```

Comment:
Promotion has more strict definition

Reply:
You are right, strictly it is more strict, but this conversions are safe, so I see no reason not to allow them.

Reply:
Renamed to `isSafeIntegralCast` to make it clear it is not only promotion